benches/fraud.rs

```diff
@@ -0,0 +1,399 @@
+//!     FROM
+//!         transactions t1 LEFT JOIN  demographics t2 ON t1.cc_num = t2.cc_num
+//! )
+//! ```
+
+mod mimalloc;
+
+use anyhow::Result;
+use clap::Parser;
+use crossbeam::channel::bounded;
+use csv::Reader as CsvReader;
+use dbsp::{
+    algebra::F64,
+    operator::{
+        time_series::{OrdPartitionedIndexedZSet, RelOffset, RelRange},
+        Avg, FilterMap,
+    },
+    CollectionHandle, DBSPHandle, OrdIndexedZSet, Runtime, Stream,
+};
+use itertools::Itertools;
+use mimalloc::MiMalloc;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use size_of::SizeOf;
+use std::{
+    hash::Hash,
+    io::{stdin, Read},
+    thread::spawn,
+};
+use time::{Date, Instant, PrimitiveDateTime};
+
+#[global_allocator]
+static ALLOC: MiMalloc = MiMalloc;
+
+const DEFAULT_BATCH_SIZE: &'static str = "10000";
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, SizeOf)]
+struct QueryResult {
+    // day: Weekday,
+    // age: u32,
+    // distance: u32,
+    // trans_diff: u32,
+    avg_spend_pw: Option<F64>,
+    avg_spend_pm: Option<F64>,
+    trans_freq_24: u32,
+    category: u32,
+    amt: F64,
+    state: u32,
+    job: u32,
+    unix_time: i32,
+    city_pop: u32,
+    merchant: u32,
+    is_fraud: u32,
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deserialize, SizeOf)]
+struct Demographics {
+    cc_num: F64,
+    first: u32,
+    gender: u32,
+    street: u32,
+    city: u32,
+    state: u32,
+    zip: u32,
+    lat: F64,
+    long: F64,
+    city_pop: u32,
+    job: u32,
+    #[size_of(skip)]
+    dob: Date,
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deserialize, SizeOf)]
+struct Transaction {
+    #[size_of(skip)]
+    #[serde(deserialize_with = "primitive_date_time_from_str")]
+    trans_date_trans_time: PrimitiveDateTime,
+    cc_num: F64,
+    merchant: u32,
+    category: u32,
+    amt: F64,
+    trans_num: String,
+    unix_time: i32,
+    merch_lat: F64,
+    merch_long: F64,
+    is_fraud: u32,
+}
+
+fn primitive_date_time_from_str<'de, D: Deserializer<'de>>(
+    d: D,
+) -> Result<PrimitiveDateTime, D::Error> {
+    let s: String = Deserialize::deserialize(d)?;
+
+    let format = time::macros::format_description!("[year]-[month]-[day] [hour]:[minute]:[second]");
+
+    match time::PrimitiveDateTime::parse(&s, &format) {
+        Ok(o) => Ok(o),
+        Err(err) => Err(D::Error::custom(err)),
+    }
+}
+
+#[derive(Debug, Clone, Parser)]
+struct Args {
+    #[clap(long)]
+    workers: usize,
+
+    #[clap(long, default_value = DEFAULT_BATCH_SIZE)]
+    batch_size: usize,
+
+    #[clap(long, default_value = "benches/fraud_data/demographics.csv")]
+    demographics: String,
+
+    #[clap(long, default_value = "benches/fraud_data/transactions.csv")]
+    transactions: String,
+
+    #[clap(long)]
+    stdin: bool,
+
+    #[doc(hidden)]
+    #[clap(long = "bench", hide = true)]
+    __bench: bool,
+}
+
+type Weight = i32;
+
+struct Fraud {
```

Comment:
I don't like the implication that DBSP is part of a Fraud.
Why don't you call this FraudBenchmarkHandle?